Upgradable Games

halgari opened the pull request:
This PR does the following: 

* Adds `LocatorIDs` onto each loadout. These are opaque strings that each game locator (for Stem, Gog, etc. ) can use to tie back to a game build of some sort
* Adds methods for getting game files, human-friendly version numbers, etc. from the LocatorIDs
* Removes the `Game Files` mod from the loadouts instead these files are injected into the synchronizer before it starts processing the diffs
* Rewrites most of `ALoadoutSynchronizer` to use massively less memory allocation, and to support injecting game files
* Adds CLI commands for changing the game versions on a loadout
* Adds diagnostics for files that are not backed up, and that are not deployable
* Adds support for Stardew Valley and Steam/Gog. For other games we'll need to add in the version mappings

Known limitations: 
* No way to back up game files currently. This will be added next in a CLI command and later into the UI
* No way to use Steam/GOG as a data source for files. This feature may wait until next sprint
* Version numbers in the loadout creation screen are stubbed out. They don't make sense there so much now that we can technically create a loadout for any game version
* There's no migration routine yet for updating old loadouts to this new format. It's trivial to write, but not part of this PR.
* The Preview changes dialog isn't updating. I'm not sure why, I'm sure it's simple, but it's late here and I'm not going to figure out tonight. I plan on fixing this issue before merging this PR.

halgari commented on extern/SMAPI:
I switched it now to 4.1.7, so it should be fixed.

halgari commented on src/Abstractions/NexusMods.Abstractions.GameLocators/GameInstallation.cs:
```diff
@@ -99,6 +94,6 @@ public override int GetHashCode()
     public override string ToString()
     {
-        return $"{Game.Name} {GameMetadataId} v{Version} ({Store.Value})";
+        return $"{Game.Name} {GameMetadataId} ({Store.Value})";
```
It will, I'll get coding on that today

halgari commented on src/Abstractions/NexusMods.Abstractions.Games.FileHashes/Models/GogBuild.cs:
```diff
@@ -30,7 +30,7 @@ public partial class GogBuild : IModelDefinition
     /// The version string of the GOG build.
     /// </summary>
-    public static readonly StringAttribute VersionName = new(Namespace, nameof(Version));
+    public static readonly StringAttribute VersionName = new(Namespace, nameof(Version)) { IsIndexed = true };
```
Index removal and addition is automatic, so no it won't require a migration.

Al12rs replied:
Nice 👍

halgari commented on src/Abstractions/NexusMods.Abstractions.Games.FileHashes/Models/VersionDefinition.cs:
```diff
@@ -0,0 +1,50 @@
+    /// </summary>
+    public static readonly GameIdAttribute GameId = new(Namespace, nameof(GameId)) { IsIndexed = true };
+    
+    /// <summary>
+    /// The gog build ids for this version
+    /// </summary>
+    public static readonly StringsAttribute GOG = new(Namespace, nameof(GOG)) { IsIndexed = true };
+    
+    /// <summary>
+    /// The associated Steam ManifestIDs for this version
+    /// </summary>
+    public static readonly StringsAttribute Steam = new(Namespace, nameof(Steam)) { IsIndexed = true };
```
No, the StringsAttribute, being multi-cardinality is inherently optional. [] is the same as not existing in this case.

halgari commented on tests/NexusMods.DataModel.Synchronizer.Tests/GeneralLoadoutManagementTests.SynchronizerIntegrationTests.verified.md:
```diff
@@ -243,10 +340,20 @@ The loadouts have been deleted and the game folder should be back to its initial
 | Path | Hash | Size |
 | --- | --- | --- |
-| {Game, bin/originalGameFile.txt} | 0x673E3C493921A2D5 | 12 B |
+| {Game, bin/newFileInGameFolderA.txt} | 0x3FB1DBAC894B6380 | 25 B |
+| {Game, bin/newFileInGameFolderA.txt} | 0x3FB1DBAC894B6380 | 25 B |
+| {Game, bin/newFileInGameFolderA.txt} | 0x3FB1DBAC894B6380 | 25 B |
+| {Game, bin/newFileInGameFolderA.txt} | 0x3FB1DBAC894B6380 | 25 B |
```
Hrm, you're right, it does look wrong, I'll take a look

halgari replied:
Wow, this took 4 hours to debug, but I finally found it, I was attaching some disk state data to the wrong entity, so the same file kept getting created over and over. Fixed.

halgari commented on src/Abstractions/NexusMods.Abstractions.Loadouts.Synchronizers/ALoadoutSynchronizer.cs:
```diff
@@ -660,12 +760,14 @@ private async Task ActionIngestFromDisk(SyncActionGroupings<SyncTreeNode> groupi
+        // Make sure the file hashes are up to date
+        await _fileHashService.GetFileHashesDb();
         return await ReindexState(gameInstallation, Connection);
     }
```
It still is needed, but I changed the name, it's not rescanning the game files, it's rescanning the files on-disk that are managed by the synchronizer.

Al12rs commented on tests/NexusMods.DataModel.Synchronizer.Tests/GeneralLoadoutManagementTests.cs:
These tests have the concept of initial state, which we don't have anymore. 
They were supposed to have `"bin/originalGameFile.txt"` as part of the initial state, but now it is just getting discarded on loadout creation. 

We talked yesterday that we should put any unrecognized files into Overrides on initial management of the game. With that we could have these tests make behave more like they were supposed to. 

Do you want to implement putting unrecognized files in override on game manage as a separate PR or as part of this one?

halgari replied:
reworked these tests and it makes more sense now, the older behavior wasn't quite right

Al12rs commented on tests/NexusMods.DataModel.Synchronizer.Tests/GeneralModManagementTests.cs:
Same as above, also counts for `GeneralFileManagementTests`

halgari commented on src/Abstractions/NexusMods.Abstractions.Loadouts.Synchronizers/SyncTreeNode.cs:
```diff
@@ -8,39 +8,96 @@
+/// <summary>
+/// A node in the synchronization tree. 24 bytes for each part, and two ushorts for the other fields. So 76 bytes total.
+/// </summary>
+public record struct SyncTreeNode
```
agreed, and in a future change, I plan on changing this data structure again to be more of a table-like format Where we can do bulk operations without having to iterate over the keys in a dictionary.

halgari commented on src/NexusMods.Games.FileHashes/Emitters/NoWayToSourceFilesOnDisk.cs:
```diff
@@ -0,0 +1,36 @@
+{
+    public async IAsyncEnumerable<Diagnostic> Diagnose(Loadout.ReadOnly loadout, CancellationToken cancellationToken)
+    {
+        var syncronizer = loadout.InstallationInstance.GetGame().Synchronizer;
+        var syncTree = await syncronizer.BuildSyncTree(loadout);
+        syncronizer.ProcessSyncTree(syncTree);
```
Agreed, that will take a update to every diagnostic in the system, (probably we'll start passing in a context struct) so I'll do that in a future PR

halgari commented on src/NexusMods.Games.FileHashes/FileHashesService.cs:
```diff
@@ -201,6 +208,173 @@ public async ValueTask<IDb> GetFileHashesDb()
+    private VersionDefinition.ReadOnly GetGameVersionDefinition(GameInstallation installation, IEnumerable<string> locatorMetadata)
+    {
+        VersionDefinition.ReadOnly versionDefinition;
+        if (installation.Store == GameStore.GOG)
```
Agreed, it's a tad strange with the way the dispatch works, but it would be good to break that out on the next store being added

Al12rs reviewed (changes requested):
I tried doing some basic tests with a Stardew installation:
Adding a new file in the game folder causes Apply to fail due to exception:

Steps:
- Manage clean SDV game
- Close app
- Add a new text file in the game folder (I used `test/newfile.txt`)
- Open the app (this will run an ingest on startup)
- Open preview and notice the new file is marked as something to be removed
- Which shouldn't be correct
- Attempt to apply -> get exception:

```
System.ArgumentException: An item with the same key has already been added. Key: {Game}/test/newfile.txt
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector)
   at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.ReindexState(GameInstallation installation, IConnection connection, ITransaction tx) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 985
   at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.ReindexState(GameInstallation installation, IConnection connection) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 961
   at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.BuildSyncTree(ReadOnly loadout) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 355
   at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.Synchronize(ReadOnly loadout) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 762
   at NexusMods.DataModel.Synchronizer.SynchronizerService.Synchronize(LoadoutId loadoutId) in C:\Dev\Repos\NexusMods.App\src\NexusMods.DataModel\Synchronizer\SynchronizerService.cs:line 66
   at NexusMods.App.UI.LeftMenu.Items.ApplyControlViewModel.<Apply>b__28_0() in C:\Dev\Repos\NexusMods.App\src\NexusMods.App.UI\LeftMenu\Items\ApplyControl\ApplyControlViewModel.cs:line 115
   at NexusMods.App.UI.LeftMenu.Items.ApplyControlViewModel.Apply() in C:\Dev\Repos\NexusMods.App\src\NexusMods.App.UI\LeftMenu\Items\ApplyControl\ApplyControlViewModel.cs:line 113
   at NexusMods.App.UI.LeftMenu.Items.ApplyControlViewModel.<>c__DisplayClass27_0.<<-ctor>b__0>d.MoveNext() in C:\Dev\Repos\NexusMods.App\src\NexusMods.App.UI\LeftMenu\Items\ApplyControl\ApplyControlViewModel.cs:line 55
``` 

EDIT:
I just tested and starting with a completely clean SDV, you are asked to Apply when you create the first loadout, which is a regression. Applying caused a bunch of exceptions, but here the Apply button then got hidden. 

A lot of exceptions from `NoWayToSourceFilesOnDisk` diagnostic going through the synchronizer:
```
System.ArgumentException: An item with the same key has already been added. Key: {Game}/api-ms-win-core-console-l1-1-0.dll
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector)
   at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.ReindexState(GameInstallation installation, IConnection connection, ITransaction tx) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 985
   at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.ReindexState(GameInstallation installation, IConnection connection) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 961
   at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.BuildSyncTree(ReadOnly loadout) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 355
   at NexusMods.Games.FileHashes.Emitters.NoWayToSourceFilesOnDisk.Diagnose(ReadOnly loadout, CancellationToken cancellationToken)+MoveNext() in C:\Dev\Repos\NexusMods.App\src\NexusMods.Games.FileHashes\Emitters\NoWayToSourceFilesOnDisk.cs:line 16
   at NexusMods.Games.FileHashes.Emitters.NoWayToSourceFilesOnDisk.Diagnose(ReadOnly loadout, CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at NexusMods.DataModel.Diagnostics.DiagnosticManager.<>c__DisplayClass8_1.<<GetLoadoutDiagnostics>b__0>d.MoveNext() in C:\Dev\Repos\NexusMods.App\src\NexusMods.DataModel\Diagnostics\DiagnosticManager.cs:line 92
--- End of stack trace from previous location ---
   at NexusMods.DataModel.Diagnostics.DiagnosticManager.<>c__DisplayClass8_1.<<GetLoadoutDiagnostics>b__0>d.MoveNext() in C:\Dev\Repos\NexusMods.App\src\NexusMods.DataModel\Diagnostics\DiagnosticManager.cs:line 92
```

halgari commented on src/Abstractions/NexusMods.Abstractions.Loadouts.Synchronizers/ALoadoutSynchronizer.cs:
```diff
@@ -660,12 +760,14 @@ private async Task ActionIngestFromDisk(SyncActionGroupings<SyncTreeNode> groupi
     {
+        // Make sure the file hashes are up to date
+        await _fileHashService.GetFileHashesDb();
         return await ReindexState(gameInstallation, Connection);
```
Removed the dead code

erri120 commented on tests/NexusMods.DataModel.SchemaVersions.Tests/MigrationSpecificTests/TestsFor_0002_RemoveGameFiles.cs:
```diff
@@ -0,0 +1,39 @@
+/// Test the migration to remove game files and add LocatorIds and GameVersions to Loadouts.
+/// </summary>
+/// <param name="provider"></param>
+public class TestsFor_0002_RemoveGameFiles(ITestOutputHelper helper) : ALegacyDatabaseTest(helper)
```
I've used `_0002_` in #2604 and `_0003_` in #2610.

erri120 commented on src/NexusMods.DataModel.SchemaVersions/Migrations/_0002_NexusCollectionItem.cs:
```diff
@@ -15,8 +15,9 @@ internal class _0002_NexusCollectionItem : ITransactionalMigration
-    public void Prepare(IDb db)
+    public async Task Prepare(IDb db)
     {
+        await Task.Yield();
```
It's a hack for async methods. Alternative is removing `async` from the method and simply returning `Task.CompletedTask`.

erri120 commented on src/Games/NexusMods.Games.StardewValley/Emitters/SMAPIGameVersionDiagnosticEmitter.cs:
```diff
@@ -40,7 +40,7 @@ public async IAsyncEnumerable<Diagnostic> Diagnose(Loadout.ReadOnly loadout, [En
 
         // var gameVersion = SimplifyVersion(new Version("1.5.6.22018"));
-        var gameVersion = new SemanticVersion(loadout.InstallationInstance.Version);
+        var gameVersion = new SemanticVersion(loadout.GameVersion);
```
For the Stardew Valley diagnostic emitters, we should use the assembly file version of the `Stardew Valley.dll` file.

halgari replied:
we can add that in, in a future PR, this keeps the existing behavior

Al12rs reviewed (changes requested):
Adding a new file in the game folder causes errors:
```
System.ArgumentException: An item with the same key has already been added. Key: {Game}/Test/NewFile.txt
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector)
   at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.ReindexState(GameInstallation installation, IConnection connection, ITransaction tx) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 1047
   at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.ReindexState(GameInstallation installation, IConnection connection) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 1023
   at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.BuildSyncTree(ReadOnly loadout) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 357
   at NexusMods.Games.FileHashes.Emitters.UndeployableLoadoutDueToMissingGameFiles.Diagnose(ReadOnly loadout, CancellationToken cancellationToken)+MoveNext() in C:\Dev\Repos\NexusMods.App\src\NexusMods.Games.FileHashes\Emitters\UndeployableLoadoutDueToMissingGameFiles.cs:line 16
   at NexusMods.Games.FileHashes.Emitters.UndeployableLoadoutDueToMissingGameFiles.Diagnose(ReadOnly loadout, CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at NexusMods.DataModel.Diagnostics.DiagnosticManager.<>c__DisplayClass8_1.<<GetLoadoutDiagnostics>b__0>d.MoveNext() in C:\Dev\Repos\NexusMods.App\src\NexusMods.DataModel\Diagnostics\DiagnosticManager.cs:line 92
--- End of stack trace from previous location ---
   at NexusMods.DataModel.Diagnostics.DiagnosticManager.<>c__DisplayClass8_1.<<GetLoadoutDiagnostics>b__0>d.MoveNext() in C:\Dev\Repos\NexusMods.App\src\NexusMods.DataModel\Diagnostics\DiagnosticManager.cs:line 92
```

halgari commented on src/Abstractions/NexusMods.Abstractions.Loadouts.Synchronizers/ALoadoutSynchronizer.cs:
```diff
@@ -681,120 +852,122 @@ protected bool HaveArchive(Hash hash)
+    protected virtual bool ShouldWin(in GamePath path, LoadoutItem.ReadOnly existing, LoadoutItem.ReadOnly replacement)
     {
-        return ValueTask.CompletedTask;
+        return replacement.Id > existing.Id;
```
they are monotomically increasing

halgari commented on src/NexusMods.App/Program.cs:
```diff
@@ -62,6 +61,9 @@ public static int Main(string[] args)
         var services = host.Services;
 
+        // Okay to do wait here, as we are in the main process thread.
+        host.StartAsync().Wait(timeout: TimeSpan.FromMinutes(5));
```
Perhaps, but no such migrations exist yet, and there isn't yet a race condition here. We'll keep an eye on it and rework it when needed

halgari commented on src/NexusMods.DataModel.SchemaVersions/Migrations/_0004_RemoveGameFiles.cs:
```diff
@@ -0,0 +1,107 @@
+    /// <summary>
+    /// DI Constructor
+    /// </summary>
+    public _0004_RemoveGameFiles(IFileHashesService fileHashService)
```
The data will still be backed up, hrm, but that is a possible concern for the GC. I'll write that down in an issue, I'm not sure if the GC will wipe out backed up game files here.

halgari commented:
> Adding a new file in the game folder causes errors:
> 
> ```
> System.ArgumentException: An item with the same key has already been added. Key: {Game}/Test/NewFile.txt
>    at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
>    at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
>    at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
>    at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector)
>    at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.ReindexState(GameInstallation installation, IConnection connection, ITransaction tx) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 1047
>    at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.ReindexState(GameInstallation installation, IConnection connection) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 1023
>    at NexusMods.Abstractions.Loadouts.Synchronizers.ALoadoutSynchronizer.BuildSyncTree(ReadOnly loadout) in C:\Dev\Repos\NexusMods.App\src\Abstractions\NexusMods.Abstractions.Loadouts.Synchronizers\ALoadoutSynchronizer.cs:line 357
>    at NexusMods.Games.FileHashes.Emitters.UndeployableLoadoutDueToMissingGameFiles.Diagnose(ReadOnly loadout, CancellationToken cancellationToken)+MoveNext() in C:\Dev\Repos\NexusMods.App\src\NexusMods.Games.FileHashes\Emitters\UndeployableLoadoutDueToMissingGameFiles.cs:line 16
>    at NexusMods.Games.FileHashes.Emitters.UndeployableLoadoutDueToMissingGameFiles.Diagnose(ReadOnly loadout, CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
>    at NexusMods.DataModel.Diagnostics.DiagnosticManager.<>c__DisplayClass8_1.<<GetLoadoutDiagnostics>b__0>d.MoveNext() in C:\Dev\Repos\NexusMods.App\src\NexusMods.DataModel\Diagnostics\DiagnosticManager.cs:line 92
> --- End of stack trace from previous location ---
>    at NexusMods.DataModel.Diagnostics.DiagnosticManager.<>c__DisplayClass8_1.<<GetLoadoutDiagnostics>b__0>d.MoveNext() in C:\Dev\Repos\NexusMods.App\src\NexusMods.DataModel\Diagnostics\DiagnosticManager.cs:line 92
> ```

Fixed this, the diagnostic run in parallel, so there's a high chance that when a new file shows up they both race and add the same file do the indexed state at the same time (one for each diagnostic). Added a lock to the ALoadoutSynchronizer to fix this. This should happen less often once we add the sync tree as an input to all diagnostics.

Al12rs reviewed (changes requested):
PR breaks the SMAPI installer, due to the removal of the GameFiles group:
![image](https://github.com/user-attachments/assets/d477aa01-2d6d-40d2-9138-91dce92c34d4)

https://github.com/Nexus-Mods/NexusMods.App/blob/626c53cc13f46849837c1958718c1c343b29ace7/src/Games/NexusMods.Games.StardewValley/Installers/SMAPIInstaller.cs#L157-L196

Basically SMAPI wants to rename a `"Stardew Valley.deps.json"` game file to `"StardewModdingAPI.deps.json"`. Doing this is apparently needed for SMAPI:
> (This is needed to resolve native DLLs like libSkiaSharp.)

I think we could do this by adding a Deletion of the original game file and a new file entry with the new path inside the SMAPI mod.
But we need access to the original game file to be able to create the copy inside the installer.
